lua/nvim-surround/patterns.lua

```diff
@@ -0,0 +1,135 @@
+    local b_first, b_last
+    -- Linewise search for the pattern before/on the cursor
+    for lnum = curpos[1], 1, -1 do
+        -- Get the file contents from the first line to current line
+        local cur_text = table.concat(buffer.get_lines(1, lnum - 1), "\n")
+        -- Find the character positions of the pattern in the file (before the cursor)
+        b_first, b_last = buffer_text:find(find, #cur_text + 1)
+        if b_first and b_first <= cursor_index then
+            break
+        end
+    end
+    -- If no match found, return the after one, if it exists
+    if not b_first or not b_last then
+        return a_first
+            and a_last
+            and {
+                first_pos = M.index_to_pos(a_first),
+                last_pos = M.index_to_pos(a_last),
+            }
+    end
+    -- Adjust the selection character-wise
+    local start_col, end_col = math.min(b_last, cursor_index), b_first
+    b_first, b_last = nil, nil
+    for index = start_col, end_col, -1 do
+        local c_first, c_last = buffer_text:find(find, index)
+        -- Validate if there is a current match
+        if c_last then
+            -- If no match yet or the current match is "better", use the current match
+            if
+                not (b_first and b_last) -- No match yet
+                or (b_last == c_last) -- Extending current match
+                or (cursor_index < b_first and c_first < b_first) -- Current is closer to cursor, after case
+                or (b_last < cursor_index and b_last < c_last) -- Current is closer to cursor, before case
+            then
+                b_first, b_last = c_first, c_last
+            end
+        end
+    end
+    -- If the cursor is inside the range then return it
+    if b_last and b_first and b_last >= cursor_index then
+        return {
+            first_pos = M.index_to_pos(b_first),
+            last_pos = M.index_to_pos(b_last),
+        }
+    end
+    -- Else if there's a range found after the cursor, return it
+    if a_first and a_last then
+        return {
+            first_pos = M.index_to_pos(a_first),
+            last_pos = M.index_to_pos(a_last),
+        }
+    end
+    -- Otherwise return the range found before the cursor, if one exists
+    if b_first and b_last then
+        return {
+            first_pos = M.index_to_pos(b_first),
+            last_pos = M.index_to_pos(b_last),
+        }
+    end
+end
+
+-- Finds the start and end indices for the given match groups.
+---@param offset integer The offset of the search string into the buffer.
+---@param str string The given string to match against.
+---@param pattern string The given Lua pattern to extract match groups from.
+---@return selections @The selections for the left and right delimiters.
+M.get_selections = function(offset, str, pattern)
+    -- Get the surrounding pair, and the start/end indices
+    local _, _, left_delimiter, first_index, right_delimiter, last_index = str:find(pattern)
```

Comment:
That looks problematic to me: if `str` is an empty string, `first_index` and `last_index` will always be `nil` so adding to it will fail

Reply:
That's a good point; I should be handling errors here in case of an invalid Lua pattern. With respect to your first point, `first_index` and `last_index` can potentially be empty strings instead of `nil`, but I agree that error-handling should be done to ensure that the program doesn't arbitrarily crash.

As for your second point, there are actually *four* sets of `()` required, see `:h nvim-surround.config.get_selections()`. I think it would make sense here to put a warning about it, rather than NOOP-ing?

Reply:
Yes I agree, a warning definitely makes sense here.

Reply:
Introduced this in a8a156c, if you would like to try it out. I'm going to re-read through the help docs/code to see if I can find any more improper uses of the words `surrounds`/`delimiters`